### DISCUSSION TITLE
SPIFFE+SPIRE as the communication pathway in a multi-cloud+multi-kubernetes-cluster CNPG DR setup.

### DISCUSSION THREAD

**Discussion by larssb** [cloudnative-pg/cloudnative-pg#6421] [Q&A]
Dear fellow CloudNativePG users. Merry Christmas.

We, in [Nemlig](https://nemlig.com) have been users of CloudNativePG for almost two years now - in production. Now we’re looking into expanding our Kubernetes multi-cluster, multi-cloud setup so that it is cloud dynamic ( a term we’ve coined ). Cloud Dynamic meaning: agnostic to the cloud provider where a Kubernetes cluster might be running. The network side of tying together clusters across different cloud regions and cloud providers is as light weight as possible. Failure is automatic and multi-cluster management is M A N A G E A B L E.

As we’re running Postgresql via CloudNativePG I’m looking into the different HA setups of PostgreSQL and support for a multi-cloud setup. I’m leaning towards: https://cloudnative-pg.io/documentation/1.24/architecture/#deployments-across-kubernetes-clusters < as I: 

- like the size of the failure domain in this case
- We’ll have a global load-balancer in front of the multi-cloud+multi-cluster setup. Likely: https://www.k8gb.io

With that lead-in to my questions.

Allow me to ask. I’m considering us to front Postgresql, running in a distributed topology where replica clusters are at the core of the capability to failover in a disaster recovery scenario.

In order to facilitate the network communication between the primary PostgreSQL cluster and the replica clusters I’m strongly considering to use SPIFFE+SPIRE, via an Envoy instance fronting the primary and the replica clusters ( I think we’ll have two - correct me please, if this is a bad idea ). I want to try this setup as it:

- Makes the network side of things agnostic to the CNI, PodCIDR’s … and we don’t have to think about whether overlay, underlay and so on is at play it’s still secure. As SPIFFE+SPIRE is mTLS
- We think that it’s very likely that we’ll have to use workload identities ( SPIFFE+SPIRE ) for other workloads than postgreSQL. So we need it anyways and it would therefore be great to re-use the technology for a PostgreSQL HA/DR setup
- No tunnels is involved as communication will be over the public Internet. Even though Tailscale and even okay open source wireguard based solutions are out there …. they all often involve some kind of need to exit and enter the different clusters through special worker nodes in the cluster acting as network traffic gateway nodes

---

THE QUESTIONS :smile:

- Have anyone tried fronting CNPG PostgreSQL with SPIFFE+SPIRE?
- What about network latency. Is the “lack” of dedicated connections between different clouds or inner cloud availability zones a problem in the distributed replica HA/DR PostgreSQL setup? I would think the extra round-trips in a tunnel related setup with worker node gateway instances will also add latency anyways and then it’ll be traffic over the Public Internet anyways

Thank you very much for any info and answer I’m given here. It’ll be very much appreciated.

If I need to elaborate further on any of the above please let me know and I’ll try to the best of my abilities.

**Comment by JustinLex** [comment]
In postgres itself, I believe authentication with both x509 SVIDs and JWT SVIDs is now supported in postgresql 18, using either client certificates or oauth authentication:

https://www.postgresql.org/docs/current/auth-cert.html

https://www.enterprisedb.com/blog/preview-postgresql-18s-oauth2-authentication-1-explore-how-it-works

I think the next step towards bringing SPIFFE to CNPG would be to write a pull request that adds these authentication options to CNPG and configuring them to use SPIFFE/SPIRE. 

With that in place, I believe users should then be able to configure SPIFFE identity authorization using `Cluster.spec.bootstrap.initdb.owner`.

**Comment by mfreeman451** [comment]
We are interested in this as well, but we are currently tied to PG 16 as we're using timescaledb. Not sure if these efforts created in 18 could be backported?

**Comment by ardentperf** [comment]
I have POC'd CNPG with SPIFFE/SPIRE.  Will see if I can share something

**Reply by mfreeman451** [reply]
@ardentperf any updates here?

**Reply by ardentperf** [reply]
@mfreeman451 https://github.com/cloudnative-pg/cloudnative-pg/discussions/6421#discussioncomment-17994994

**Comment by hausteen** [comment]
Any updates here on this topic? I'd love to see cnpg support SPIFFE/SPIRE natively.

**Reply by gbartolini** [reply]
Theoretically, with the new support for TLS client certificates and database role, we should have done an important step towards that direction. Would anyone be interested in doing deeper analysis on this?

**Reply by mfreeman451** [reply]
@gbartolini can you expand on this a little bit? why would the new support for TLS client certs be an important step etc?

**Comment by ardentperf** [comment]
Here's a full example of doing postgres and spiffe integration with docker. this particular example is interesting because it doesn't require any client awareness at all - the client connects with sslmode=disable and a dummy username and password. Two Envoy sidecar proxies handle the mTLS tunnel, mutual identity verification using X.509 SVIDs, and - via a Wasm network filter - JWT SVID fetch from SPIRE, username rewriting, and credential injection into the Postgres authentication exchange, all transparently on behalf of the application.

On the server side, it require a custom PAM module with `spire-agent` and a small shell wrapper.

https://github.com/geico/database-spiffe-auth-examples/tree/main/postgres

I haven't tried to port this to CNPG yet. I'd like to see if we can use an extension container image to make the spire-agent and shell wrapper available in the postgres container.

I think right now pgbouncer only supports pam on an all-or-nothing basis; i don't think `auth_hba_file` supports it, so if it's enabled then every connection needs to use it.

**Reply by ardentperf** [reply]
fyi - discussion on pgbouncer & pam https://github.com/pgbouncer/pgbouncer/issues/1546

**Reply by hausteen** [reply]
Have you looked at Istio? It seems like this is what you are trying to achieve.
